### PR TITLE
delay closeEmojiPicker event listener

### PR DESCRIPTION
**Pull request by DanC5**
# Submit a pull request

## CLA

- [ ] I have signed the [Stream CLA](https://docs.google.com/forms/d/e/1FAIpQLScFKsKkAJI7mhCr7K9rEIOpqIDThrWxuvxnwUq2XkHyG154vQ/viewform) (required).
- [ ] The code changes follow best practices
- [ ] Code changes are tested (add some information if not applicable)

### CODE REVIEW

File: src/components/MessageInput/hooks/messageInput.js [hunk 1]
```diff
@@ -354,7 +354,10 @@ export default function useMessageInput(props) {
       type: 'setEmojiPickerIsOpen',
       value: true,
     });
-    document.addEventListener('click', closeEmojiPicker, false);
+
+    setTimeout(() =>
```
Comment (vishalnarkhede) [inline src/components/MessageInput/hooks/messageInput.js:358]: Can we add a comment here as to why we need timeout?

Comment (DanC5) [reply src/components/MessageInput/hooks/messageInput.js:359]: good call, added!

Comment (vishalnarkhede) [reply src/components/MessageInput/hooks/messageInput.js:359]: How about useEffect?

```js
useEffect(() => {
    if (emojiPickerIsOpen) {
        document.addEventListener('click', closeEmojiPicker, false);
    }

    return () => {
        document.removeEventListener('click', closeEmojiPicker, false);
    }
}, [emojiPickerIsOpen])
```

**Review by ath92 — COMMENTED** [review]
I tried this locally, but for me the current changes only allow me to open the emoji picker once (without the changes in this PR, it doesn't open at all).

I tried @vishalnarkhede 's solution with `useEffect`. I like it because it ensures that the event handler always gets cleaned up, which the current code doesn't. You can see this happening when you run a development server; as you make changes to the code while the picker is open, multiple event handlers are added to the document, because they're only cleaned up when you close the picker. Unfortunately, Vishal's solution doesn't seem to solve the race for me, and the picker doesn't open with only `useEffect`.

I'm still not quite sure how this used to work, but I think I found a clean solution:
```
  const closeEmojiPicker = useCallback(
    (e) => {
      if (
        emojiPickerRef.current &&
        !emojiPickerRef.current.contains(e.target)
      ) {
        dispatch({
          type: 'setEmojiPickerIsOpen',
          value: false,
        });
      }
    },
    [emojiPickerRef],
  );

  const openEmojiPicker = useCallback((e) => {
    dispatch({
      type: 'setEmojiPickerIsOpen',
      value: true,
    });
    // Prevent event from bubbling to document, so the close handler is never called for this event
    e.stopPropagation();
  }, []);

  useEffect(() => {
    if (state.emojiPickerIsOpen) {
      document.addEventListener('click', closeEmojiPicker, false);
    }
    return () => {
      document.removeEventListener('click', closeEmojiPicker, false);
    };
  }, [closeEmojiPicker, state.emojiPickerIsOpen]);
```

**Review by ath92 — APPROVED** [review]

**Pull request merged**